magma/topological_sort.py

```diff
@@ -0,0 +1,241 @@
+        for dependency in dependencies:
+            if dependency not in non_duplicated_dependencies:
+                non_duplicated_dependencies.append(dependency)
+        dictionary[name_of_dependencies_key] = non_duplicated_dependencies
+        #TODO: note -- im working under the assumption that because of the limitations
+        # of the schema, the dependencies will be of the correct type. Must I include
+        # a check that each dependency is in fact a name of another metaworkflow?
+        #(...probably. :/ )
+
+        # check for self-dependencies
+        new_dependencies = dictionary[name_of_dependencies_key] #repetitive, but just for readability
+        dictionary_name = dictionary[name_of_step_key]
+        # remove from this list
+        #TODO: should I throw exception instead? I think it's fine to just remove bc it's easy
+        new_dependencies = list(filter(lambda element: element != dictionary_name, new_dependencies))
+
+    return list_of_dicts_copy
+
+def generate_ordered_step_name_list(list_of_dicts, name_of_step_key):
+    """
+    Based on a list of dictionaries (representing a list of steps) with a "name" key
+    for each dictionary, return a list of the names of each dictionary with
+    indices corresponding to the indices of the dictionaries themselves (same order).
+    """
+    names = []
+    for dictionary in list_of_dicts:
+        names.append(dictionary[name_of_step_key])
+    return names
+    #TODO: in test, check that it is always in the same order
+
+def define_forward_dependencies(list_of_dicts, name_of_step_key, name_of_dependencies_key):
+    """
+    Build directed graph by "reversing" dependencies TODO: redo comment
+    """
+    names = generate_ordered_step_name_list(list_of_dicts, name_of_step_key)
+
+    for dictionary in list_of_dicts:
+        current_dependencies = dictionary[name_of_dependencies_key]
+        current_dict_name = dictionary[name_of_step_key]
+
+        # go through each step this current step is dependent on
+        # and add "step_after" attribute
+        # (dependencies are "origin", "source", or "progenitor" steps)
+        for dependency in current_dependencies:
+            # isolate the index of the dependency using names list
+            #TODO: this matches to the first occurence of dependency within the array 
+            idx = names.index(dependency)
+            dependency_step_dict = list_of_dicts[idx]
+
+            #TODO: consider helper fxn? but not necessary
+            #TODO: rename this attribute
+            if not ("steps_after" in dependency_step_dict):
+                dependency_step_dict["steps_after"] = []
+
+            dependency_step_dict["steps_after"].append(current_dict_name)
+
+def find_index_with_given_step_name(steps_with_dependencies_list, name_of_step_key, name):
+    for index, step in enumerate(steps_with_dependencies_list):
+        if step[name_of_step_key] == name:
+            return index, step
```

Comment:
Is it possible to not find the given name? This will return `None`, which may cause problems.